src/main/java/net/logstash/logback/util/ReusableByteBuffers.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.util;
+
+import java.lang.ref.Reference;
+import java.lang.ref.SoftReference;
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+/**
+ * A pool of {@link ReusableByteBuffer}.
+ * 
+ * <p>The pool is unbounded and can hold as many buffers as needed. Buffers are kept in the pool
```

Comment:
Unbounded is a little scary to me.

Under what conditions would this grow "too big" ?


Also, what are your thoughts on adding some monitoring capability for this?  That could be a future enhancement, but we could plan for it now.

Reply:
Yeah... say that if you have 200 concurrent threads requesting a buffer exactly at the same time, then you will have 200 buffers in the pool. But this is quite unlikely to happen... especially if the Encoder is relatively fast. You would have had exactly the same if buffers were kept in a ThreadLocal...
By using a ConcurrentLinkedQueue we limit the amount of buffers in the pool to what the maximum concurrency actually requires - which is not the case with a ThreadLocal.

As it is for the moment, entries will be removed from the pool only when the JVM needs some more memory...

If the Encoder is used by one of our async appender - then we know it will always be from the same thread and the pool size will always be one.
If the Encoder is used by other kinds of appender - then the pool can grow up to the maximum number of threads logging at exactly the same time - not more.

With this in mind, I'm not convinced we should think about alternative strategies to actively reduce/shrink the buffer other than the SoftReference... your opinion?

Reply:
> If the Encoder is used by one of our async appender - then we know it will always be from the same thread and the pool size will always be one.
If the Encoder is used by other kinds of appender - then the pool can grow up to the maximum number of threads logging at exactly the same time - not more.

Got it.  Thanks for the explanation.  This makes me feel better about it being "unbounded".  Essentially, while it is technically true that it is unbounded... in practice it should never grow more than the number of threads accessing it concurrently.

I think this behavior is worth mentioning in the javadoc for the paranoid (like me) that get scared when they see "unbounded".  ;)

> With this in mind, I'm not convinced we should think about alternative strategies to actively reduce/shrink the buffer other than the SoftReference... your opinion?

I agree.  The current solution is good.